src/runtime_src/hip/core/event.h

```diff
@@ -124,22 +128,58 @@ class kernel_start : public command
   bool wait() override;
 };
 
+// copy command for copying data from/to host buffer of type void*
 class copy_buffer : public command
 {
 public:
-  copy_buffer(std::shared_ptr<stream> s, xclBOSyncDirection direction, std::shared_ptr<memory> buf, void* ptr, size_t size, size_t offset);
+  copy_buffer(std::shared_ptr<stream> s, xclBOSyncDirection direction, std::shared_ptr<memory> buf, void* host_buf, size_t size, size_t offset)
+      : command(command::type::buffer_copy, std::move(s)), cdirection(direction), buffer(std::move(buf)), host_buffer(host_buf), copy_size(size), dev_offset(offset)
+  {}
   bool submit() override;
   bool wait() override;
 
-private:
-  xclBOSyncDirection cdirection;
-  std::shared_ptr<memory> buffer;
-  void* host_ptr;
+protected:
+  xclBOSyncDirection cdirection; // copy direction
+  std::shared_ptr<memory> buffer; // device buffer
+  void* host_buffer; // host buffer copy source/destination
   size_t copy_size;
   size_t dev_offset; // offset for device memory
   std::future<void> handle;
 };
 
+// copy command for copying data from a source only bost buffer of type std::vector<uint8|uint16|uint32>
+template<class T>
+class copy_from_host_buffer_command : public copy_buffer
+{
+public:
+  copy_from_host_buffer_command(std::shared_ptr<stream> s, xclBOSyncDirection direction, std::shared_ptr<memory> buf, std::vector<T>& host_vec, size_t size, size_t offset)
```

Comment:
```suggestion
  copy_from_host_buffer_command(std::shared_ptr<stream> s, xclBOSyncDirection direction, std::shared_ptr<memory> buf, std::vector<T>&& host_vec, size_t size, size_t offset)
```
This must take an rvalue reference that can then be moved into the data member without incurring copy.

Reply:
Done.